### DISCUSSION TITLE
One variable for different Connection (MysqlConnection, SqliteConnection)

### DISCUSSION THREAD

**Discussion by linrongbin16** [diesel-rs/diesel#2986] [Q&A] [answered]
Hi diesel team, I'm quite new to Rust. Trying to create database connection with diesel-rs.

I want to create different connection type (sometimes mysql, sometimes sqlite) by command parameters, but provide the same database API for my app.

That means, after initialize the database, my app code will use the one connection for all db logics.

Here's part of my code:
```rust
use diesel::Connecction;
use diesel::mysql::MysqlConnection;
use diesel::sqlite::SqliteConnection;

let engine = "mysql";
let mysql_url = "mysql://username:password@localhost:3306/test";
let sqlite_url = "sqlite://sqlite.db";
let connection : dyn Connection = if engine == "mysql" {
  MysqlConnection::establish(mysql_url).unwrap()
} else {
  SqliteConnection::establish(sqlite_url).unwrap()
}
```

Here's the compiler error:
```
error[E0191]: the value of the associated types `Backend` (from trait `Connection`), `TransactionManager` (from trait `Connection`) must be specified
  --> src/quant/common/persistence/database.rs:11:25
   |
11 |     connection: Box<dyn Connection>,
   |                         ^^^^^^^^^^ help: specify the associated types: `Connection<Backend = Type, TransactionManager = Type>`

```

Is it possible to do this ?

**Comment by weiznich** [comment, marked as answer]
That's not supported by diesel, as you can already see by looking at the corresponding error message.

Edit: To add some more information about this use case:

What is not supported by diesel is creating **one** connection trait object that could contain connections to different database backends. We need to know which backend is used at compile time, as otherwise we cannot verify that the constructed query is valid for this backend at compile time. This means there are two solutions to your problem:

* Make the connection choice a compile time option via a feature flag on your crate. In this case you want to conditionally define a `type DbConnection = SqliteConnection/MysqlConnection` somewhere in your crate root.
* Write code in a generic way. I only recommend doing that for people with really advanced knowledge about rusts trait system and that try to rewrite reusable library crates. Otherwise you will run in really complex trait bounds here, so probably don't even try that if your are new to rust.
